### ISSUE TITLE
is MPU9250 works well with arduino uno ?

### ISSUE THREAD

**Issue by ahmd-gamal** [kriswiner/MPU-9250#411]
hi kriswiner ,
is MPU9250 works well with Arduino Uno ?

**Comment by kriswiner** [comment]
yes, but MPU9250 is 3V3 only and the fusion rate will be very low

**Comment by ahmd-gamal** [comment]
the data stays the same even if I move the sensor
ax = 0.00 ay = 0.00 az = 0.00 mg
gx = 0.00 gy = 0.00 gz = 0.00 deg/s
mx = -470 my = -120 mz = -125 mG
q0 = 1.00 qx = 0.00 qy = 0.00 qz = 0.00
Yaw, Pitch, Roll: -13.80, 0.00, 0.00
rate = 630.19 Hz

**Comment by kriswiner** [comment]
You are not receiving any data.

Maybe you don;t have it connected properly.

**Comment by ahmd-gamal** [comment]
i was scanning my i2c
Scanning...
I2C device found at address 0x68  !
done

and WHO_AM_I_MPU9250 is working fine ,, 

MPU9250 I AM 71 I should be 71

and my connection is 
 VDD ---------------------- 5V (there is a voltage regulator (5 to 3.3 v) built-in  
 SDA ----------------------- A4
 SCL ----------------------- A5
 GND ---------------------- GND

**Comment by kriswiner** [comment]
Then maybe don;t have the interrupt woking?

**Comment by ahmd-gamal** [comment]
sorry, I do not understand what you mean. can you give me more details please

**Comment by kriswiner** [comment]
Send me a link to the exact sketch you are using.

**Comment by ahmd-gamal** [comment]
//

**Comment by kriswiner** [comment]
I asked for a link.

What are ares and gres values?

**Comment by ahmd-gamal** [comment]
I was made some modifications to the code because it doesn't work well with Arduino oun . the modifications are : 
- add MadgwickQuaternionUpdate and MahonyQuaternionUpdate function from quaternionFilters file  to the file that linked below because it was not declared on the scope 
- removing all lcd commands because there is no memory in arduino .

https://github.com/kriswiner/MPU9250/blob/master/MPU9250BasicAHRS.ino

What are ares and gres values?
Gscale = GFS_250DPS
Ascale = AFS_2G
so, 
aRes = 2.0/32768.0;
gRes = 250.0/32768.0;

**Comment by kriswiner** [comment]
Are these the values printed out on the serial monitor for ares and gres?

**Comment by ahmd-gamal** [comment]
I think the values didn't print at all .so this is the value printed out 👍   

MPU9250 I AM 71 I should be 71
MPU9250 is online...
x-axis self test: acceleration trim within : -100.0% of factory value
y-axis self test: acceleration trim within : -100.0% of factory value
z-axis self test: acceleration trim within : -100.0% of factory value
x-axis self test: gyration trim within : -100.0% of factory value
y-axis self test: gyration trim within : -100.0% of factory value
z-axis self test: gyration trim within : -100.0% of factory value
MPU9250 initialized for active data mode....
AK8963 I AM FF I should be 48
AK8963 initialized for active data mode....
X-Axis sensitivity adjustment value 0.50
Y-Axis sensitivity adjustment value 0.50
Z-Axis sensitivity adjustment value 0.50
ax = 0.00 ay = 0.00 az = 0.00 mg
gx = 0.00 gy = 0.00 gz = 0.00 deg/s
mx = -470 my = -120 mz = -125 mG
q0 = 1.00 qx = 0.00 qy = 0.00 qz = 0.00
Yaw, Pitch, Roll: -13.80, 0.00, 0.00
rate = 0.16 Hz
ax = 0.00 ay = 0.00 az = 0.00 mg
gx = 0.00 gy = 0.00 gz = 0.00 deg/s
mx = -470 my = -120 mz = -125 mG
q0 = 1.00 qx = 0.00 qy = 0.00 qz = 0.00
Yaw, Pitch, Roll: -13.80, 0.00, 0.00
rate = 647.51 Hz

**Comment by kriswiner** [comment]
Looks like the MPU9250 is not out of sleep mode, which is the mode on
initial power on.

The sensor is providing no data.

**Comment by ahmd-gamal** [comment]
maybe the problem here 
(AK8963 I AM FF I should be 48 )
what do you think ?

**Comment by kriswiner** [comment]
Your MPU9250 is not "on", you need to start the PLL and put it into normal
run mode.

**Comment by luizamanfrini** [comment]
Hey Kris, I think I am having the same problem here. How do I put it into normal run mode?

**Comment by kriswiner** [comment]
Consult the data sheet and register map or look at the github sketches

// wake up device
  writeByte(MPU9250_ADDRESS, PWR_MGMT_1, 0x00); // Clear sleep mode bit
(6), enable all sensors
  delay(100); // Wait for all registers to reset

 // get stable time source
  writeByte(MPU9250_ADDRESS, PWR_MGMT_1, 0x01);  // Auto select clock
source to be PLL gyroscope reference if ready else
  delay(200);

**Comment by Mustafakyuz** [comment]
Hi Kris, firstly thanks for codes and informations. I have same problem. I'm using arduino uno and mpu9250. And I added MadgwickQuaternionUpdate  function from quaternionFilters file (MPU9250BasicAHRS.ino). But I have a error that 'q' was not declared in this scope. I dont know what I changing in code. Please help me.
this is code which I work:


[basic_ahrstest1.zip](https://github.com/kriswiner/MPU9250/files/6644759/basic_ahrstest1.zip)


when  I try compare this code, ı have error which :

C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino: In function 'void MadgwickQuaternionUpdate(float, float, float, float, float, float, float, float, float)':
basic_ahrstest1:181:24: error: 'q' was not declared in this scope
             float q1 = q[0], q2 = q[1], q3 = q[2], q4 = q[3];   // short name local variable for readability
                        ^
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:181:24: note: suggested alternative: 'q1'
             float q1 = q[0], q2 = q[1], q3 = q[2], q4 = q[3];   // short name local variable for readability
                        ^
                        q1
basic_ahrstest1:195:33: error: 'q2' was not declared in this scope
             float _2q2 = 2.0f * q2;
                                 ^~
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:195:33: note: suggested alternative: 's2'
             float _2q2 = 2.0f * q2;
                                 ^~
                                 s2
basic_ahrstest1:196:33: error: 'q3' was not declared in this scope
             float _2q3 = 2.0f * q3;
                                 ^~
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:196:33: note: suggested alternative: 's3'
             float _2q3 = 2.0f * q3;
                                 ^~
                                 s3
basic_ahrstest1:197:33: error: 'q4' was not declared in this scope
             float _2q4 = 2.0f * q4;
                                 ^~
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:197:33: note: suggested alternative: 's4'
             float _2q4 = 2.0f * q4;
                                 ^~
                                 s4
basic_ahrstest1:252:61: error: 'beta' was not declared in this scope
             qDot1 = 0.5f * (-q2 * gx - q3 * gy - q4 * gz) - beta * s1;
                                                             ^~~~
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:252:61: note: suggested alternative: 'gets'
             qDot1 = 0.5f * (-q2 * gx - q3 * gy - q4 * gz) - beta * s1;
                                                             ^~~~
                                                             gets
basic_ahrstest1:258:27: error: 'deltat' was not declared in this scope
             q1 += qDot1 * deltat;
                           ^~~~~~
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:258:27: note: suggested alternative: 'delay'
             q1 += qDot1 * deltat;
                           ^~~~~~
                           delay
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino: In function 'void loop()':
basic_ahrstest1:500:3: error: 'MahonyQuaternionUpdate' was not declared in this scope
   MahonyQuaternionUpdate(ax, ay, az, gx*PI/180.0f, gy*PI/180.0f, gz*PI/180.0f, my, mx, mz);
   ^~~~~~~~~~~~~~~~~~~~~~
C:\Users\makyu\Desktop\basic_ahrstest1\basic_ahrstest1.ino:500:3: note: suggested alternative: 'MadgwickQuaternionUpdate'
   MahonyQuaternionUpdate(ax, ay, az, gx*PI/180.0f, gy*PI/180.0f, gz*PI/180.0f, my, mx, mz);
   ^~~~~~~~~~~~~~~~~~~~~~
   MadgwickQuaternionUpdate
exit status 1
'q' was not declared in this scope

**Comment by kriswiner** [comment]
Add the QuaternionFilter.ino to your sketch tab; put the two files in the
same sketch folder.

**Comment by MoradSayed** [comment]
@Mustafakyuz 
Can you pass the working code, if you still have it 
Thanks in advance.
